retriever/lib/provenance.py

```diff
@@ -0,0 +1,92 @@
+    for package in packages:
+        if package:
+            package_name, version = str(package).split(' ')
+            details[package_name] = version
+    return details
+
+
+def commit_info(dataset):
+    """
+    Generate info for a particular commit.
+    """
+    info = {'packages': package_details(),
+            'time': datetime.now(timezone.utc).strftime("%m/%d/%Y, %H:%M:%S"),
+            'version': dataset.version}
+    return info
+
+
+def commit(dataset, commit_message='', path='.', quiet=False):
+    """
+    Commit dataset to a zipped file.
+    """
+    if type(dataset) == str:
+        # if dataset is not a dataset script object find the right script
+        dataset = [script for script in datasets() if script.name == dataset][0]
+    paths_to_zip = {'script': dataset._file,
+                    'raw_data': []}
+    raw_dir = os.path.join(HOME_DIR, 'raw_data')
+    data_exists = False
+    if not quiet:
+        print("Committing dataset {}".format(dataset.name))
+    try:
+        if dataset.name not in os.listdir(raw_dir):
+            engine = choose_engine({'command': 'download', 'path': './', 'sub_dir': ""})
+            dataset.download(engine=engine, debug=quiet)
+            data_exists = True
+
+        elif dataset.name in os.listdir(raw_dir):
+            data_exists = True
+
+        if data_exists:
+            for root, _, files in os.walk(os.path.join(raw_dir, dataset.name)):
+                for file in files:
+                    paths_to_zip['raw_data'].append(os.path.join(root, file))
+
+            info = commit_info(dataset)
+            info['commit_message'] = commit_message
+            info['script_name'] = os.path.basename(dataset._file)
+
+            if os.path.exists(os.path.join(HOME_DIR, 'raw_data', dataset.name)):
+                info['md5_dataset'] = getmd5(os.path.join(HOME_DIR, 'raw_data', dataset.name), 'dir', encoding=ENCODING)
+            info['md5_script'] = getmd5(dataset._file)
+            with ZipFile(os.path.join(path, '{}-{}{}.zip'.format(dataset.name,
+                                                                 info['md5_dataset'][:3],
+                                                                 info['md5_script'][:3])), 'w') as zipped:
+                zipped.write(paths_to_zip['script'],
+                             os.path.join('script', os.path.basename(paths_to_zip['script'])))
+
+                for data_file in paths_to_zip['raw_data']:
+                    zipped.write(data_file, data_file.replace(raw_dir, ""))
+                with open('metadata.json', 'w') as json_file:
+                    json.dump(info, json_file, sort_keys=True, indent=4)
+                zipped.write(os.path.abspath(json_file.name), 'metadata.json')
+                os.remove('metadata.json')
+        if not quiet:
+            print("Successfully committed.")
+    except Exception as e:
+        print("Dataset could not be committed.")
+        raise (e)
```

Comment:
Avoid raising, in case we have to use other programs, the raise will interfere with the python interpreter. Let's print the error e and use return.

Reply:
Okay :+1: